src/Transform/AffLin.hs

```diff
@@ -260,21 +260,29 @@ affLinDiscards [] = return []
 affLinProg :: Prog -> AffLinM Prog
 affLinProg (ProgData y cs) =
   pure (ProgData y (mapCtors affLinTp cs))
-affLinProg (ProgFun x as tm tp) =
-  let as' = mapParams affLinTp as
-      tp' = affLinTp tp
-  in pure (\tm' -> ProgFun x as' tm' tp') <*> alBinds as' (affLin tm)
-affLinProg (ProgExtern x ps tp) =
-  pure (ProgExtern x (map affLinTp ps) (affLinTp tp))
+affLinProg (ProgFun x tp tm) =
+  -- Top-level arrows are not transformed
+  let
+    (ptps, rtp) = splitArrows tp
+    tp' = joinArrows (fmap affLinTp ptps) (affLinTp rtp)
+    (ps, rtm) = splitLams tm
```

Comment:
Is it ok if `ptps` and `ps` are not the same length?

Reply:
I thought they would be guaranteed to be the same length, but is there an edge case?

```
% cat bug.ppl
define f = \x: Bool -> Bool. x;

f (\b. b) True

% ./perplc -c  bug.ppl
data Bool = False | True;

define f : <Bool -> Bool, ()> -> Bool -> Bool = \ x : <Bool -> Bool, ()>. x;

f <\ b : Bool. b, ()> True

% ./perplc -z bug.ppl
perplc: misshaped tensor index
CallStack (from HasCallStack):
  error, called at ./Util/Tensor.hs:102:20 in main:Util.Tensor
```

Weirdly, it generates an FGG without error.

Reply:
Here's how it looks in master:

```
% ./perplc -c bug.ppl
data Bool = False | True;

define f : <Bool -> Bool, ()> -> Bool -> Bool = \ x : <Bool -> Bool, ()>. \ _f1 : Bool. (let <x0, _> = x in x0) _f1;

f <\ b : Bool. b, ()> True

% ./perplc -z bug.ppl
[0.0, 1.0]%
```